### PR TITLE
Accept block expressions

### PR DESCRIPTION
**Pull request by amscanne**
These are a specific form of blocks (e.g. `{ ... } `) which are
terminated by an expression that has no semi-value. They are accepted
anywhere that an expression is accepted.

When assigned to a map, block expressions will automatically propagate
the map semantics to the final statement in the expression.

An example of a block expression:
```
BEGIN { $y = 0; @x = { $z = 1; $z }; @a = @x ? { let $c = $y; $c+5 } : { let $b = 4; $b }; exit() }
```

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by jordalgo** [comment]
Just as a reference, which I think @amscanne mentioned in another thread, block expressions are value expressions in Rust as well: https://doc.rust-lang.org/reference/expressions/block-expr.html

Overall, I think I can get behind this proposal even though recently I was advocating for printing bare variable/global expressions (maybe we can print them by default if they are the last expression in a probe?! 🤔 ). I also can't really think of a good reason not to allow this.

File: man/adoc/bpftrace.adoc [hunk 1]
```diff
@@ -891,6 +891,21 @@ All valid identifiers can be used as `name`.
 
 The data type of a variable is automatically determined during first assignment and cannot be changed afterwards.
 
+==== Block Expressions
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:894]: I think this should go either above comments: https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#comments

Or inside https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#operators-and-expressions

File: src/parser.yy [hunk 2]
```diff
@@ -526,13 +527,16 @@ tuple_access_expr:
                 postfix_expr DOT INT      { $$ = driver.ctx.make_node<ast::FieldAccess>($1, $3, @3); }
                 ;
 
-
+block_expr:
+                "{" stmt_list expr "}" { $$ = driver.ctx.make_node<ast::Block>(std::move($2), $3, @$); }
```
Comment (jordalgo) [inline src/parser.yy:531]: Should we also allow the final expression to be a call as well? Rust allows this.

**Review by jordalgo — COMMENTED** [review]
I think we should also add a few parser tests and runtime tests but otherwise LGTM

**Review by viktormalik — COMMENTED** [review]
Just two rewordings for the docs, otherwise looks good.

**Comment by amscanne** [comment]
Thanks for the reviews! I'll clean up the wording and add a few more tests. Are there any specific edge cases that come to mind? (Do you think we need runtime tests specifically? I'm not sure what edges you'd want to cover there, although I do feel it's reasonable to add quite a few more unit tests.)

File: man/adoc/bpftrace.adoc [hunk 3]
```diff
@@ -891,6 +891,21 @@ All valid identifiers can be used as `name`.
...
+
+```
+let $a = {
+  let $b = 1;
+  $b
+}
+// $a is 1
+```
+
+This can be used anywhere a regular expression can be used.
```
[the remark this answers is not in the document]
Comment (viktormalik) [reply man/adoc/bpftrace.adoc:907]: Sure, that sounds good.

File: CHANGELOG.md [hunk 4]
```diff
@@ -29,6 +29,8 @@ and this project adheres to
   - [#3811](https://github.com/bpftrace/bpftrace/pull/3811)
 - Add support for LLVM 20
   - [#3841](https://github.com/bpftrace/bpftrace/pull/3841)
+- Added simple block expressions
```
[the remark this answers is not in the document]
Comment (amscanne) [reply CHANGELOG.md]: Done

File: man/adoc/bpftrace.adoc [hunk 5]
```diff
@@ -891,6 +891,21 @@ All valid identifiers can be used as `name`.
...
+==== Block Expressions
+
+A block can be used as expression, as long as the last line of the block is an
+expression with no trailing semi-colon.
+
+```
+let $a = {
+  let $b = 1;
+  $b
+}
```
[the remark this answers is not in the document]
Comment (amscanne) [reply man/adoc/bpftrace.adoc]: Yes, it makes more sense that way. Done.

File: man/adoc/bpftrace.adoc [hunk 6]
```diff
@@ -891,6 +891,21 @@ All valid identifiers can be used as `name`.
 
 The data type of a variable is automatically determined during first assignment and cannot be changed afterwards.
 
+==== Block Expressions
+
+A block can be used as expression, as long as the last line of the block is an
```
[the remark this answers is not in the document]
Comment (amscanne) [reply man/adoc/bpftrace.adoc]: Done

File: man/adoc/bpftrace.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (amscanne) [reply man/adoc/bpftrace.adoc]: Fixed

File: man/adoc/bpftrace.adoc (hunk 1 above)
Comment (amscanne) [reply man/adoc/bpftrace.adoc]: Moved to the operators and expressions section

**Review by viktormalik — APPROVED** [review]
LGTM.

Nice addition!

File: src/ast/ast.cpp [hunk 7]
```diff
@@ -184,7 +184,19 @@ AssignMapStatement::AssignMapStatement(Diagnostics &d,
...
+  // they are being assigned.
+  Expression *value = expr;
+  while (true) {
+    Block *block = dynamic_cast<Block *>(value);
+    if (block == nullptr) {
+      break;
+    }
+    value = block->expr; // Must be non-null if expression.
+  };
+  value->map = map;
```
Comment (jordalgo) [inline src/ast/ast.cpp:199]: Just an FYI: this seems to further call out the async behavior of our map printing e.g.
```sh
# bpftrace -e 'BEGIN { @a = { @a = 2; print(@a); 1 }; exit(); }'
Attaching 1 probe...
@a: 1

@a: 1
```

**Comment by jordalgo** [comment]
One small clandy-tidy error, which TBH I find a bit silly, that needs fixing - otherwise I think this is good to go.

**Comment by jordalgo** [comment]
Just re-gen codegen tests and we'll get this merged

**Comment by amscanne** [comment]
Ah, because of the rebase. Done!
